File: AiTrainerIOS/AiTrainerCore/Sources/AiTrainerCore/QuestionRepository.swift
import Foundation

public final class QuestionRepository {
    private let bundle: Bundle
    private let documentsDirectory: URL
    private var loaded = false
    private var loadErrorInternal: String?
    private var questionsInternal: [Question] = []
    private var duplicateIdMapInternal: [String: String] = [:]
    private var newQuestionIdsInternal: [String] = []
    private var byIdInternal: [String: Question] = [:]

    public init(bundle: Bundle = .main, documentsDirectory: URL? = nil) {
        self.bundle = bundle
        self.documentsDirectory = documentsDirectory ?? FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
    }

    public var loadError: String? {
        ensureLoaded()
        return loadErrorInternal
    }

    public var isReady: Bool { loadError == nil }

    public var questions: [Question] {
        ensureLoaded()
        return questionsInternal
    }

    public var duplicateIdMap: [String: String] {
        ensureLoaded()
        return duplicateIdMapInternal
    }

    public var newQuestionIds: [String] {
        ensureLoaded()
        return newQuestionIdsInternal
    }

    public var usesImportedBank: Bool {
        importedFileURL().flatMap { FileManager.default.fileExists(atPath: $0.path) } ?? false
    }

    public func findById(_ id: String) -> Question? {
        ensureLoaded()
        return byIdInternal[canonicalId(id)]
    }

    public func canonicalId(_ id: String) -> String {
        duplicateIdMapInternal[id] ?? id
    }

    public func allIds() -> [String] { questions.map(\.id) }

    public func byType(_ type: String) -> [Question] {
        questions.filter { $0.type == type }
    }

    public func questionsForBanks(_ enabledBanks: Set<BankKind>) -> [Question] {
        let types = Set(enabledBanks.map(\.type))
        if types.isEmpty { return [] }
        return questions.filter { types.contains($0.type) }
    }

    public func activeIds(_ enabledBanks: Set<BankKind>) -> Set<String> {
        Set(questionsForBanks(enabledBanks).map(\.id))
    }

    public func countForKind(_ kind: BankKind) -> Int {
        questions.count { $0.type == kind.type }
    }

    public func reload() {
        loaded = false
        loadErrorInternal = nil
        ensureLoaded()
    }

    public func bankInfo(enabledBanks: Set<BankKind>) -> QuestionBankInfo {
        let fullSingle = countForKind(.single)
        let fullJudge = countForKind(.judge)
        let active = questionsForBanks(enabledBanks)
        return QuestionBankInfo(
            total: active.count,
            singleCount: active.count { $0.type == BankKind.single.type },
            judgeCount: active.count { $0.type == BankKind.judge.type },
            sourceLabel: usesImportedBank ? "导入题库" : "内置题库",
            canRestoreBuiltIn: usesImportedBank,
            fullSingleCount: fullSingle,
            fullJudgeCount: fullJudge
        )
    }

    private func ensureLoaded() {
        if loaded { return }
        loaded = true
        do {
            duplicateIdMapInternal = try loadJSON(named: "duplicate_id_map", fallbackURL: nil) ?? [:]
            newQuestionIdsInternal = try loadJSON(named: "new_question_ids", fallbackURL: nil) ?? []
            if let imported = readImportedQuestions() {
                questionsInternal = imported
            } else if let builtIn: [Question] = try loadJSON(named: "questions", fallbackURL: nil) {
                questionsInternal = builtIn
            } else {
                throw NSError(domain: "QuestionRepository", code: 1, userInfo: [NSLocalizedDescriptionKey: "无法加载内置题库"])
            }
            byIdInternal = Dictionary(uniqueKeysWithValues: questionsInternal.map { ($0.id, $0) })
        } catch {
            loadErrorInternal = error.localizedDescription
            questionsInternal = []
            byIdInternal = [:]
        }
    }

    private func importedFileURL() -> URL? {
        documentsDirectory.appendingPathComponent(AppConfig.importedFileName)
    }

    private func readImportedQuestions() -> [Question]? {
        guard let url = importedFileURL(), FileManager.default.fileExists(atPath: url.path) else { return nil }
        let data = try? Data(contentsOf: url)
        return data.flatMap { try? JSONDecoder().decode([Question].self, from: $0) }
    }

    private func loadJSON<T: Decodable>(named name: String, fallbackURL: URL?) throws -> T? {
        if let url = bundle.url(forResource: name, withExtension: "json") {
            let data = try Data(contentsOf: url)
            return try JSONDecoder().decode(T.self, from: data)
        }
        if let fallbackURL {
            let data = try Data(contentsOf: fallbackURL)
            return try JSONDecoder().decode(T.self, from: data)
        }
        return nil
    }
}

File: AiTrainer/app/src/main/java/com/aitrainer/practice/data/QuestionRepository.kt
package com.aitrainer.practice.data

import android.content.Context
import android.net.Uri
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import java.io.File
import java.io.InputStream

class QuestionRepository(context: Context) {

    private val gson = Gson()
    private val appContext = context.applicationContext
    private val customFile get() = File(appContext.filesDir, IMPORTED_FILE)

    private var loaded = false
    private var loadErrorInternal: String? = null
    private var questionsInternal: List<Question> = emptyList()
    private var duplicateIdMapInternal: Map<String, String> = emptyMap()
    private var newQuestionIdsInternal: List<String> = emptyList()
    private var byIdInternal: Map<String, Question> = emptyMap()

    val loadError: String? get() {
        ensureLoaded()
        return loadErrorInternal
    }

    val isReady: Boolean get() = loadError == null

    val questions: List<Question> get() {
        ensureLoaded()
        return questionsInternal
    }

    val duplicateIdMap: Map<String, String> get() {
        ensureLoaded()
        return duplicateIdMapInternal
    }

    val newQuestionIds: List<String> get() {
        ensureLoaded()
        return newQuestionIdsInternal
    }

    val usesImportedBank: Boolean
        get() = customFile.exists()

    fun findById(id: String): Question? {
        ensureLoaded()
        return byIdInternal[canonicalId(id)]
    }

    fun canonicalId(id: String): String = duplicateIdMap[id] ?: id

    fun canonicalIds(ids: List<String>): List<String> {
        val valid = questions.map { it.id }.toSet()
        val out = mutableListOf<String>()
        val seen = mutableSetOf<String>()
        ids.forEach { raw ->
            val cid = canonicalId(raw)
            if (valid.contains(cid) && seen.add(cid)) out.add(cid)
        }
        return out
    }

    fun allIds(): List<String> = questions.map { it.id }

    fun byType(type: String): List<Question> = questions.filter { it.type == type }

    fun questionsForBanks(enabledBanks: Set<BankKind>): List<Question> {
        val types = enabledBanks.map { it.type }.toSet()
        if (types.isEmpty()) return emptyList()
        return questions.filter { it.type in types }
    }

    fun activeIds(enabledBanks: Set<BankKind>): Set<String> =
        questionsForBanks(enabledBanks).map { it.id }.toSet()

    fun countForKind(kind: BankKind): Int = questions.count { it.type == kind.type }

    fun reload() {
        loaded = false
        loadErrorInternal = null
        ensureLoaded()
    }

    fun importFromUri(uri: Uri, target: BankKind, policy: DuplicatePolicy = DuplicatePolicy.SKIP): Result<MergeStats> {
        return runCatching {
            val parsed = appContext.contentResolver.openInputStream(uri)?.use { stream ->
                parseQuestions(stream)
            } ?: error("无法读取文件")
            require(parsed.isNotEmpty()) { "题库为空" }
            mergeQuestionsIntoBank(parsed, target, policy).getOrThrow()
        }
    }

    fun mergeQuestionsIntoBank(
        incoming: List<Question>,
        target: BankKind,
        policy: DuplicatePolicy = DuplicatePolicy.SKIP,
    ): Result<MergeStats> {
        return runCatching {
            require(incoming.isNotEmpty()) { "没有可导入的题目" }
            val base = readBaseQuestions()
            val preserved = base.filter { it.type != target.type }
            val targetPool = base.filter { it.type == target.type }
            val prepared = incoming.map { it.adaptToBank(target) }
            val outcome = mergeIntoPool(
                existing = targetPool,
                incoming = prepared,
                policy = policy,
                reservedIds = preserved.map { it.id }.toSet(),
            )
            writeQuestions(preserved + outcome.questions)
            reload()
            require(loadErrorInternal == null) { loadErrorInternal ?: "导入后加载失败" }
            outcome.stats
        }
    }

    fun mergeQuestions(
        incoming: List<Question>,
        policy: DuplicatePolicy = DuplicatePolicy.SKIP,
    ): Result<MergeStats> {
        return runCatching {
            require(incoming.isNotEmpty()) { "没有可导入的题目" }
            val outcome = mergeIntoPool(readBaseQuestions(), incoming, policy)
            writeQuestions(outcome.questions)
            reload()
            require(loadErrorInternal == null) { loadErrorInternal ?: "导入后加载失败" }
            outcome.stats
        }
    }

    private data class MergeOutcome(val questions: List<Question>, val stats: MergeStats)

    private fun mergeIntoPool(
        existing: List<Question>,
        incoming: List<Question>,
        policy: DuplicatePolicy,
        reservedIds: Set<String> = emptySet(),
    ): MergeOutcome {
        val map = existing.associateBy { it.id }.toMutableMap()
        val usedIds = (existing.map { it.id } + reservedIds).toMutableSet()
        var added = 0
        var updated = 0
        var skipped = 0

        incoming.forEach { question ->
            when {
                question.id in map && policy == DuplicatePolicy.SKIP -> skipped++
                question.id in map && policy == DuplicatePolicy.REPLACE -> {
                    map[question.id] = question
                    updated++
                }
                question.id in usedIds && policy == DuplicatePolicy.RENAME -> {
                    val newId = renameId(question.id, usedIds)
                    val renamed = question.copy(id = newId, tag = newId)
                    map[newId] = renamed
                    usedIds.add(newId)
                    added++
                }
                question.id in usedIds -> skipped++
                else -> {
                    map[question.id] = question
                    usedIds.add(question.id)
                    added++
                }
            }
        }

        return MergeOutcome(map.values.toList(), MergeStats(added = added, updated = updated, skipped = skipped))
    }

    private fun Question.adaptToBank(target: BankKind): Question = when (target) {
        BankKind.SINGLE -> if (type == target.type) this else copy(type = target.type)
        BankKind.JUDGE -> copy(
            type = target.type,
            options = listOf("正确", "错误"),
            answer = when {
                answer in JUDGE_ANSWERS -> answer
                answer.contains("正确") || answer.trim() in setOf("对", "√", "T", "Y") -> "正确"
                answer.contains("错误") || answer.trim() in setOf("错", "×", "F", "N") -> "错误"
                else -> answer
            },
        )
    }

    fun clearImportedBank(): Boolean {
        if (!customFile.exists()) return false
        customFile.delete()
        reload()
        return loadErrorInternal == null
    }

    fun bankInfo(enabledBanks: Set<BankKind> = BankKind.entries.toSet()): QuestionBankInfo {
        ensureLoaded()
        val qs = questionsInternal
        val active = questionsForBanks(enabledBanks)
        val fullSingle = qs.count { it.type == BankKind.SINGLE.type }
        val fullJudge = qs.count { it.type == BankKind.JUDGE.type }
        return QuestionBankInfo(
            total = active.size,
            singleCount = active.count { it.type == BankKind.SINGLE.type },
            judgeCount = active.count { it.type == BankKind.JUDGE.type },
            sourceLabel = if (usesImportedBank) "导入题库" else "内置题库",
            canRestoreBuiltIn = usesImportedBank,
            fullSingleCount = fullSingle,
            fullJudgeCount = fullJudge,
        )
    }

    private fun ensureLoaded() {
        if (loaded) return
        loaded = true
        runCatching {
            questionsInternal = readQuestions()
            duplicateIdMapInternal = readAssetJson("duplicate_id_map.json", object : TypeToken<Map<String, String>>() {}.type)
            newQuestionIdsInternal = readAssetJson("new_question_ids.json", object : TypeToken<List<String>>() {}.type)
            byIdInternal = questionsInternal.associateBy { it.id }
        }.onFailure { error ->
            AppLog.e("Failed to load question bank", error)
            loadErrorInternal = "题库加载失败（${error.message ?: "未知错误"}）"
            questionsInternal = emptyList()
            duplicateIdMapInternal = emptyMap()
            newQuestionIdsInternal = emptyList()
            byIdInternal = emptyMap()
        }
    }

    private fun readQuestions(): List<Question> = readBaseQuestions()

    private fun readBaseQuestions(): List<Question> {
        val reader = if (customFile.exists()) {
            customFile.bufferedReader()
        } else {
            appContext.assets.open("questions.json").bufferedReader()
        }
        return reader.use { parseQuestions(it) }
    }

    private fun writeQuestions(questions: List<Question>) {
        customFile.outputStream().use { out ->
            out.writer().use { writer ->
                gson.toJson(questions, writer)
            }
        }
    }

    private fun renameId(baseId: String, existing: Set<String>): String {
        var suffix = 1
        while ("$baseId-$suffix" in existing) suffix++
        return "$baseId-$suffix"
    }

    private fun parseQuestions(input: java.io.Reader): List<Question> {
        val type = object : TypeToken<List<Question>>() {}.type
        return gson.fromJson<List<Question>>(input, type).map { it.normalizedForStorage() }
    }

    private fun Question.normalizedForStorage(): Question = copy(
        answerExpl = answerExpl?.trim()?.takeIf { it.isNotEmpty() },
        optionExpls = optionExpls?.filterValues { it.isNotBlank() }?.takeIf { it.isNotEmpty() },
    )

    private fun parseQuestions(input: InputStream): List<Question> =
        input.bufferedReader().use { parseQuestions(it) }

    private fun <T> readAssetJson(name: String, type: java.lang.reflect.Type): T =
        appContext.assets.open(name).bufferedReader().use { gson.fromJson(it, type) }

    companion object {
        private const val IMPORTED_FILE = "imported_questions.json"
        private val JUDGE_ANSWERS = setOf("正确", "错误")
    }
}
